docs(python): Add security warning in LazyFrame.deserialize() docstring

douglas-raillard-arm opened the pull request:
Add indication that LazyFrame.deserialize() might execute arbitrary code coming from the deserialized data.

Fixes #14623

stinodego commented on py-polars/polars/lazyframe/frame.py:
```diff
@@ -610,6 +610,11 @@ def deserialize(cls, source: str | Path | IOBase) -> Self:
         """
         Read a logical plan from a JSON file to construct a LazyFrame.
 
+        .. warning::
+            This function uses :mod:`pickle` under some circumstances, and as
+            such inherits the security implications. Deserializing can execute
+            arbitrary code so it should only be attempted on trusted data.
```
Could you add this as a `Warnings` section rather than an admonition? The section should be below Parameters / above See Also. See https://numpydoc.readthedocs.io/en/latest/format.html

ritchie46 replied:
We should elaborate here. If we put a big fat security warning we should explain under which conditions. Polars only uses pickle if the logical plan was serialized with python UDFs. In most queries Polars won't use pickle.

douglas-raillard-arm replied:
@ritchie46 is that relevant for deserialize() though ? From the security perspective, what matters is whether you trust the data. If so, it does not matter whether it's using pickle. However, if you don't trust the data, it's very unlikely you would trust a promise that this JSON does not contain the UDF. As such, if the warning applies to your case, you'd have to assume it contains a malicious bit of pickle. So unless there is a way for the user to prevent loading pickle at the loss of some feature (e.g. a `deserialize(..., forbid_pickle=True)` parameter), the conditions of production don't really matter.

ritchie46 replied:
It gives you more information on when we call pickle, I think that's important. You can evenly verify that the JSON doesn't have any UDFs as it is human readable.

ritchie46 replied:
And for what it's worth, I can see us adding a forbid pickle.

stinodego reviewed (changes requested):
Thanks, the text looks good. One note about formatting.

I think this also applies to `Expr.deserialize`, right?

douglas-raillard-arm commented:
Updated the PR with:
* Use of warnings section instead of Sphinx warning directive
* Added information on the circumstances leading to having pickle blobs in the `serialize()` JSON

ritchie46 commented on py-polars/polars/expr/expr.py:
```diff
@@ -344,6 +344,13 @@ def deserialize(cls, source: str | Path | IOBase) -> Self:
+            This function uses :mod:`pickle` under some circumstances, and as
+            such inherits the security implications. Deserializing can execute
+            arbitrary code so it should only be attempted on trusted data.
+            Currently, pickle will be used when serializing UDF.
```
```suggestion
            pickle is only used when the logical plan contains python UDFs.
```

ritchie46 commented on py-polars/polars/lazyframe/frame.py:
```diff
@@ -617,6 +617,14 @@ def deserialize(cls, source: str | Path | IOBase) -> Self:
+            This function uses :mod:`pickle` under some circumstances, and as
+            such inherits the security implications. Deserializing can execute
+            arbitrary code so it should only be attempted on trusted data.
+            Currently, pickle will be used when serializing UDF.
```
```suggestion
            pickle is only used when the logical plan contains python UDFs.
```

ritchie46 commented:
Thanks!
